### PR TITLE
Ads

### LINKED ISSUE #46: Feature : Ads
**Is your feature request related to a problem? Please describe.**
As a B I want to add many ads to my organization in order to register C applications.

**Describe the solution you'd like**
The applications must contain: 
- A title
- A description
- A price without taxes
- Taxes
- The all taxes included prices must be recorded but not be displayed to the user.

**Additional context**
The applications will display a percentage of completion, a C score, and many pictures in the future but must not be implemented now.

### PR DESCRIPTION
**Pull request by H-L**
#### :tophat: What? Why?
Adding real estate Ads to the application.

#### :pushpin: Related Issues
- Fixes #49 
- Fixes #46

#### :clipboard: Subtasks
- [x] Add/modify seeds
- [x] Add tests

### CODE REVIEW

File: app/controllers/ads_controller.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  # POST /ads
+  # POST /ads.json
+  def create # rubocop:disable Metrics/MethodLength
+    authorize organization, policy_class: AdPolicy
+    @ad = organization.ads.build(ad_params)
+
+    respond_to do |format|
+      if @ad.save
+        format.html { redirect_to organization_ad_path(organization, @ad), notice: "Ad was successfully created." }
```
Comment (armandfardeau) [inline app/controllers/ads_controller.rb]:
```suggestion
        format.html { redirect_to organization_ad_path(organization, @ad), notice: t(".created") }
```

File: app/controllers/ads_controller.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      end
+    end
+  end
+
+  # PATCH/PUT /ads/1
+  # PATCH/PUT /ads/1.json
+  def update
+    respond_to do |format|
+      if @ad.update(ad_params)
+        format.html { redirect_to organization_ad_path(organization, @ad), notice: "Ad was successfully updated." }
```
Comment (armandfardeau) [inline app/controllers/ads_controller.rb]:
```suggestion
        format.html { redirect_to organization_ad_path(organization, @ad), notice: t(".updated") }
```

File: app/controllers/ads_controller.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  # rubocop:enable Metrics/AbcSize
+
+  # DELETE /ads/1
+  # DELETE /ads/1.json
+  def destroy
+    @ad.destroy
+    respond_to do |format|
+      format.html { redirect_to organization_ads_url, notice: "Ad was successfully destroyed." }
```
Comment (armandfardeau) [inline app/controllers/ads_controller.rb]:
```suggestion
      format.html { redirect_to organization_ads_url, notice: t(".destroyed") }
```

File: app/models/ad.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+# Model for Ads
+class Ad < ApplicationRecord
+  belongs_to :organization
+
+  resourcify
+
+  validates :title, presence: true
+  validates :price_ec, presence: true
```
Comment (armandfardeau) [inline app/models/ad.rb]:
```suggestion
  validates :price_ec, presence: true, numericality: { greater_than_or_equal_to: 0 }
```

File: app/models/ad.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  validates :charges, presence: true
```
Comment (armandfardeau) [inline app/models/ad.rb]:
```suggestion
  validates :charges, presence: true, numericality: { greater_than_or_equal_to: 0 }
```

File: app/views/ads/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<h1><%= t(".title") %></h1>
+
+<%= render "form", ad: @ad, organization: @organization %>
+
+<%= link_to "Show", organization_ad_path(organization_id: @organization.id, id: @ad.id) %> |
```
Comment (armandfardeau) [inline app/views/ads/edit.html.erb]:
```suggestion
<%= link_to t(".show"), organization_ad_path(organization_id: @organization.id, id: @ad.id) %> |
```

File: app/views/ads/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+<%= link_to "Back", organization_ads_path(@organization) %>
```
Comment (armandfardeau) [inline app/views/ads/edit.html.erb]:
```suggestion
<%= link_to t(".back"), organization_ads_path(@organization) %>
```

File: app/views/ads/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  <tbody>
+    <% @ads.each do |ad| %>
+      <tr>
+        <td><%= ad.title %></td>
+        <td><%= ad.description %></td>
+        <td><%= ad.price_ec %></td>
+        <td><%= ad.charges %></td>
+        <td>
+          <% if policy(ad).show? %>
+            <%= link_to "Show", organization_ad_path(organization_id: @organization.id, id: ad.id) %>
```
Comment (armandfardeau) [inline app/views/ads/index.html.erb]:
```suggestion
            <%= link_to t(".show"), organization_ad_path(organization_id: @organization.id, id: ad.id) %>
```

File: app/views/ads/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+          <% end %>
+        </td>
+        <td>
+          <% if policy(ad).edit? %>
+            <%= link_to "Edit", edit_organization_ad_path(organization_id: @organization.id, id: ad.id) %>
```
Comment (armandfardeau) [inline app/views/ads/index.html.erb]:
```suggestion
            <%= link_to t(".edit"), edit_organization_ad_path(organization_id: @organization.id, id: ad.id) %>
```

File: app/views/ads/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+          <% end %>
+        </td>
+        <td>
+          <% if policy(ad).destroy? %>
+            <%= link_to "Destroy", organization_ad_path(organization_id: @organization.id, id: ad.id), method: :delete, data: { confirm: "Are you sure?" } %>
```
Comment (armandfardeau) [inline app/views/ads/index.html.erb]:
```suggestion
            <%= link_to t(".destroy"), organization_ad_path(organization_id: @organization.id, id: ad.id), method: :delete, data: { confirm: t(".confirmation") } %>
```

File: app/views/ads/index.html.erb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        </td>
+      </tr>
+    <% end %>
+  </tbody>
+</table>
+
+<br />
+
+<% if policy(@organization).edit? %>
+  <%= link_to "New Ad", new_organization_ad_path(@organization) %>
```
Comment (armandfardeau) [inline app/views/ads/index.html.erb]:
```suggestion
  <%= link_to t(".new_ad"), new_organization_ad_path(@organization) %>
```

File: app/views/ads/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<h1><%= t(".title") %></h1>
+
+<%= render "form", ad: @ad, organization: @organization %>
+
+<%= link_to "Back", organization_ads_path(@organization) %>
```
Comment (armandfardeau) [inline app/views/ads/new.html.erb]:
```suggestion
<%= link_to t(".back"), organization_ads_path(@organization) %>
```

File: app/views/ads/show.html.erb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  <%= @ad.price_ec %>
+</p>
+
+<p>
+  <strong><%= t(".charges") %></strong>
+  <%= @ad.charges %>
+</p>
+
+<% if policy(@ad).edit? %>
+  <%= link_to "Edit", edit_organization_ad_path(organization_id: @organization.id, id: @ad.id) %> |
```
Comment (armandfardeau) [inline app/views/ads/show.html.erb]:
```suggestion
  <%= link_to t(".edit"), edit_organization_ad_path(organization_id: @organization.id, id: @ad.id) %> |
```

File: app/views/ads/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+<% end %>
+
+<% if policy(@ad).destroy? %>
+  <%= link_to "Destroy", organization_ad_path(organization_id: @organization.id, id: @ad.id), method: :delete, data: { confirm: "Are you sure?" } %> |
```
Comment (armandfardeau) [inline app/views/ads/show.html.erb]:
```suggestion
  <%= link_to t(".destroy"), organization_ad_path(organization_id: @organization.id, id: @ad.id), method: :delete, data: { confirm: t(".confirmation") } %> |
```

File: app/views/ads/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+<% end %>
+
+<%= link_to "Back", organization_ads_path(@organization) %>
```
Comment (armandfardeau) [inline app/views/ads/show.html.erb]:
```suggestion
<%= link_to t(".back"), organization_ads_path(@organization) %>
```

File: app/views/organizations/show.html.erb [hunk 16]
```diff
@@ -15,6 +15,10 @@
   <%= @organization.description %>
 </p>
 
+<p>
+  <%= link_to "Organization Ads", organization_ads_path(@organization) %>
```
Comment (armandfardeau) [inline app/views/organizations/show.html.erb]:
```suggestion
  <%= link_to t(".organization_ads"), organization_ads_path(@organization) %>
```

File: app/views/organizations/show.html.erb [hunk 17]
```diff
@@ -15,6 +15,10 @@
...
+</p>
+
 <% if policy(@organization).edit? %>
   <%= link_to "Edit", edit_organization_path(@organization) %> |
```
Comment (armandfardeau) [inline app/views/organizations/show.html.erb]:
```suggestion
  <%= link_to t(".edit"), edit_organization_path(@organization) %> |
```

File: app/models/organization.rb [hunk 18]
```diff
@@ -3,15 +3,23 @@
...
 class Organization < ApplicationRecord
   belongs_to :user
+  has_many :ads
+
   resourcify
 
   validates :name, presence: true
 
-  after_create :assign_admin
+  after_create :assign_manager
```
Comment (armandfardeau) [inline app/models/organization.rb]: The hook should be after_save.
The hook should update any related ad.

File: spec/models/ad_spec.rb [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    expect(ad).not_to be_valid
+  end
+
+  describe "#calculate_price_ic" do
+    let!(:ad) { create(:ad, organization: organization) }
+
+    it "should calculate price_ic" do
+      expect(ad.price_ic).to eq(ad.price_ec + ad.charges)
+    end
+  end
```
Comment (armandfardeau) [inline spec/models/ad_spec.rb]: Check heritance of user role on ads related to an organization.

File: spec/models/organization_spec.rb [hunk 20]
```diff
@@ -20,7 +20,7 @@
     expect(organization).not_to be_valid
   end
 
-  describe "#assign_admin" do
+  describe "#assign_manager" do
     let!(:organization) { create(:organization, user: user) }
 
     it "assigns manager role to user" do
```
Comment (armandfardeau) [inline spec/models/organization_spec.rb]: Add test for #users_with_role

**Comment by H-L** [comment]
I added a few tests in the `ad_policy_spec.rb` to check if different type of user have permissions depending on their roles and organization.

**Comment by armandfardeau** [comment]
@H-L I cannot understand why test on `./spec/policies/ad_policy_spec.rb:60` fails.

This method is already tested on the user model. Could you check?

**Comment by H-L** [comment]
I don't know, actually I was using `let` in the first place, but we agreed on using `let!` instead because of some weird behavior I had on the preview tests.

Anyway, should I had a test for `propagate_to_children` method in `organization`model ?

Thanks for adding the test on `add_and_propagate_role_to` btw.

**Pull request merged**